### PR TITLE
[REEF-1891] Allow passing job submission environment variables from .…

### PR DESCRIPTION
**Pull request by jwang98052**
…Net client

* Update AvroYarnClusterJobSubmissionParameters schema for environment map
* Update Java bridge client to desterilize and set the map
* Set the map to ContainerLaunchContext
* Expose it in JobRequest and JobParameters in C#
* Serialize it in YarnREEFParamSerializer.
* Update unit test at both Java and C#
* Updated HelloREEFYarn to set it.

JIRA: [REEF-1891](https://issues.apache.org/jira/browse/REEF-1891)
This closes #

### CODE REVIEW

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 1]
```diff
@@ -241,6 +242,29 @@ public YarnSubmissionHelper setConfigurationFilePaths(final List<String> configu
   }
 
   /**
+   * Sets environment variable map.
+   * @param map
+   * @return
+   */
+  public YarnSubmissionHelper setJobSubmissionEnvMap(final Map<String, String> map) {
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: This API is unclear: Do I set the whole map, or one entry at a time?

File: lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs [hunk 2]
```diff
@@ -89,6 +92,19 @@ public JobParametersBuilder SetDriverMemory(int driverMemoryInMb)
         }
 
         /// <summary>
+        /// Adds job submission environment variable.
+        /// If the variable is already in the map, override it. 
+        /// </summary>
+        /// <param name="key"></param>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        public JobParametersBuilder AddJobSubmissionEnvVariable(string key, string value)
```
[the remark this answers is not in the document]
Comment (jwang98052) [reply lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs]: The new one will override existing one.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java [hunk 3]
```diff
@@ -121,9 +131,18 @@ public String toString() {
         ", tokenService='" + tokenService + '\'' +
         ", fileSystemUrl='" + fileSystemUrl + '\'' +
         ", jobSubmissionDirectoryPrefix='" + jobSubmissionDirectoryPrefix + '\'' +
+        envMapString() +
         '}';
   }
 
+  private String envMapString() {
```
[the remark this answers is not in the document]
Comment (jwang98052) [reply lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java]: It is just for log purpose for toString() to call. EnvMap is just part of the Avro schema. Sserialize will get everything.

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java (hunk 1 above)
Comment (jwang98052) [reply lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: Both set entire map and add individual entries are supported.

**Comment by jwang98052** [comment]
We can have an enum to define all supported keys by YARN. That will make it strong types. If Yarn ever add any new keys, we will need to update pour code to add it. 

Allow any key string is flexible, but also easy to make mistake such as typo error. 

Take entire map from client is one option, allow client to add key value pair from API is more convenient. Basically the question is shall we expose YARN API parameters directly to REEF client, or shall we have our own way to expose it.

**Comment by markusweimer** [comment]
Maybe we need constants for the known keys, but allow arbitrary keys to be used?

Also, we should call the `add...` methods `set...` as they really have those override semantics and not an `add` semantics.

**Comment by jwang98052** [comment]
"Maybe we need constants for the known keys, but allow arbitrary keys to be used." -- Sounds good. I will get existing keys defined in YARN.

**Comment by jwang98052** [comment]
@markusweimer I have addressed your review comments and pushed the change. Thanks!

File: lang/cs/Org.Apache.REEF.Client/API/JobParameters.cs [hunk 4]
```diff
@@ -81,6 +86,14 @@ public int DriverMemoryInMB
         }
 
         /// <summary>
+        /// The job submission environment variable map.
+        /// </summary>
+        public IDictionary<string, string> JobSubmissionEnvMap
+        {
+            get { return _jobSubmissionEnvMap; }
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/API/JobParameters.cs]: This returns a reference to a mutable collection, which seems dangerous. Wouldn't it be better to return a copy? Or rather: Is access to this collection as a whole needed at all?

File: lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (markusweimer) [reply lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs]: Please rename this method to `SetJobSubmissionEnvVariable` to reflect its semantics. Also, consider using `SetJobSubmissionEnvironmentVariable`.

File: lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs [hunk 5]
```diff
@@ -89,6 +92,19 @@ public JobParametersBuilder SetDriverMemory(int driverMemoryInMb)
...
+        {
+            _jobSubmissionMap[key] = value;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs]: Maybe we should add an optional out bool variable to indicate whether or not an overwrite happened?

File: lang/cs/Org.Apache.REEF.Client/API/JobRequestBuilder.cs [hunk 6]
```diff
@@ -148,6 +148,18 @@ public JobRequestBuilder SetDriverMemory(int driverMemoryInMb)
         }
 
         /// <summary>
+        /// Adds a job submission environment variable.
+        /// </summary>
+        /// <param name="key">key of the environment variable.</param>
+        /// <param name="value">Value of the environment variable.</param>
+        /// <returns></returns>
+        public JobRequestBuilder AddJobSubmissionEnvVariable(string key, string value)
```
[the remark this answers is not in the document]
Comment (markusweimer) [reply lang/cs/Org.Apache.REEF.Client/API/JobRequestBuilder.cs]: Same as above: Rename to `Set`

File: lang/cs/Org.Apache.REEF.Client/YARN/Environment.cs [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+namespace Org.Apache.REEF.Client.YARN
+{
+    /// <summary>
+    /// Default environment map keys from YARN.
+    /// </summary>
+    public enum Environment
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/Environment.cs]: `sealed`?

File: lang/cs/Org.Apache.REEF.Client/YARN/Environment.cs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+namespace Org.Apache.REEF.Client.YARN
+{
+    /// <summary>
+    /// Default environment map keys from YARN.
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/Environment.cs]: Is there reference documentation on YARN which we could link to from here?

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs [hunk 9]
```diff
@@ -91,6 +91,8 @@ private void Run()
                 .AddDriverConfiguration(driverConfig.Build())
                 .AddGlobalAssemblyForType(typeof(HelloDriverYarn))
                 .SetJobIdentifier("HelloREEF")
+                .AddJobSubmissionEnvVariable(Environment.PATH.ToString(), "value1")
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: What are these changes in the example for?

File: lang/java/reef-bridge-client/src/main/avro/JobSubmissionParameters.avsc [hunk 10]
```diff
@@ -60,6 +60,7 @@
         { "name": "securityTokenKind", "type": "string", "default": "NULL" },
         { "name": "securityTokenService", "type": "string", "default": "NULL" },
         { "name": "driverMemory", "type": "int" },
+        {"name": "envMap", "type": {"type": "map", "values": "string"}},
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/avro/JobSubmissionParameters.avsc]: Consider `environmentVariablesMap` to make the name more self-documenting.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java [hunk 11]
```diff
@@ -89,6 +92,13 @@ private YarnClusterSubmissionFromCS(final AvroYarnAppSubmissionParameters yarnAp
     this.yarnDriverStdoutFilePath = yarnClusterJobSubmissionParameters.getDriverStdoutFilePath().toString();
     this.yarnDriverStderrFilePath = yarnClusterJobSubmissionParameters.getDriverStderrFilePath().toString();
 
+    if (yarnClusterJobSubmissionParameters.getEnvMap() != null) {
+      for (Map.Entry<java.lang.CharSequence, java.lang.CharSequence> pair :
+          yarnClusterJobSubmissionParameters.getEnvMap().entrySet()) {
+        this.envMap.put(pair.getKey().toString(), pair.getValue().toString());
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java]: There must be a utility function in Apache Commons to do this :)

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java [hunk 12]
```diff
@@ -181,6 +200,13 @@ String getFileSystemUrl() {
   }
 
   /**
+   * @return The environment map.
+   */
+  Map<String, String> getEnvMap() {
+    return envMap;
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java]: This returns a reference to a mutable collection. Is this safe? Maybe return a copy?

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java [hunk 13]
```diff
@@ -241,6 +242,29 @@ public YarnSubmissionHelper setConfigurationFilePaths(final List<String> configu
...
+    return this;
+  }
+
+  /**
+   * Adds a job submission environment variable.
+   * @param key
+   * @param value
+   * @return
+   */
+  public YarnSubmissionHelper addJobSubmissionEnvVariable(final String key, final String value) {
```
Comment (markusweimer) [inline lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: rename to `setJobSubmissionEnvironmentVariable`

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java (hunk 12 above)
Comment (markusweimer) [reply lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java]: Also, consider renaming to `getEnvironmentVariablesMap` or such.

File: lang/cs/Org.Apache.REEF.Client/API/JobParameters.cs (hunk 4 above)
Comment (jwang98052) [reply lang/cs/Org.Apache.REEF.Client/API/JobParameters.cs]: Ok

File: lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs (hunk 5 above)
Comment (jwang98052) [reply lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs]: The usage is like build pattern. The user won't have chance to check the out parameter. Make the parameter like <key, value> is clean and simple. It is client's responsibility to set it. If they want to override, just override.

File: lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwang98052) [reply lang/cs/Org.Apache.REEF.Client/API/JobParametersBuilder.cs]: Done

File: lang/cs/Org.Apache.REEF.Client/YARN/Environment.cs (hunk 7 above)
Comment (jwang98052) [reply lang/cs/Org.Apache.REEF.Client/YARN/Environment.cs]: sealed is not valid for enum

File: lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs (hunk 9 above)
Comment (jwang98052) [reply lang/cs/Org.Apache.REEF.Examples.HelloREEF/HelloREEFYarn.cs]: It is to show people example and also test the code path.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java (hunk 11 above)
Comment (jwang98052) [reply lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java]: Didn't find good one. Do you know any?

File: lang/java/reef-bridge-client/src/main/avro/JobSubmissionParameters.avsc (hunk 10 above)
Comment (jwang98052) [reply lang/java/reef-bridge-client/src/main/avro/JobSubmissionParameters.avsc]: Done

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java (hunk 12 above)
Comment (jwang98052) [reply lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/YarnClusterSubmissionFromCS.java]: Done

File: lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java (hunk 13 above)
Comment (jwang98052) [reply lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/client/YarnSubmissionHelper.java]: Done

**Comment by jwang98052** [comment]
@markusweimer I have addressed your review comments. Please have another look. Thanks!

**Comment by jwang98052** [comment]
@markusweimer Can you look at it again? More future changes will touch the same files in this PL. So I would like it to be merged before starting other JIRA to avoid conflict.

**Pull request closed without merging**